iothub/device/src/Authentication/IotHubConnectionCredentials.cs

```diff
@@ -112,61 +117,188 @@ public IotHubConnectionCredentials(string iotHubConnectionString)
+                }
+
+                // This assignment resets any previously set SharedAccessSignature value. This is possible in flows where the same authentication method instance
+                // is used to reinitialize the client after close-dispose.
+                // SharedAccessSignature should be set only if it is non-null and the authentication method of the device client is
+                // not of type AuthenticationWithTokenRefresh.
+                // Setting the SAS value for an AuthenticationWithTokenRefresh authentication type will result in tokens not being renewed.
+                // This flow can be hit if the same authentication method is always used to initialize the client;
+                // as in, on disposal and reinitialization. This is because the value of the SAS token computed is stored within the authentication method,
+                // and on reinitialization the client is incorrectly identified as a fixed-sas-token-initialized client,
+                // instead of being identified as a sas-token-refresh-enabled-client.
+                SharedAccessSignature = null;
+
+                Debug.Assert(SasTokenRefresher != null);
+            }
         }
 
-        private void ExtractPropertiesFromConnectionString(string iotHubConnectionString)
+        private void SetAuthenticationModel()
         {
-            IDictionary<string, string> map = iotHubConnectionString.ToDictionary(ValuePairDelimiter, ValuePairSeparator);
-
-            HostName = GetConnectionStringValue(map, HostNamePropertyName);
-            GatewayHostName = GetConnectionStringOptionalValue(map, GatewayHostNamePropertyName);
-            DeviceId = GetConnectionStringOptionalValue(map, DeviceIdPropertyName);
-            ModuleId = GetConnectionStringOptionalValue(map, ModuleIdPropertyName);
-            SharedAccessKeyName = GetConnectionStringOptionalValue(map, SharedAccessKeyNamePropertyName);
-            SharedAccessKey = GetConnectionStringOptionalValue(map, SharedAccessKeyPropertyName);
-            SharedAccessSignature = GetConnectionStringOptionalValue(map, SharedAccessSignaturePropertyName);
+            AuthenticationModel = Certificate == null
+                ? SharedAccessKeyName == null
+                    ? AuthenticationModel.SasIndividual
+                    : AuthenticationModel.SasGrouped
+                : AuthenticationModel.X509;
         }
 
-        internal void Validate()
+        private void Validate()
         {
-            // Host name
-            if (HostName.IsNullOrWhiteSpace())
+            // IoT Hub host name
+            if (IotHubHostName.IsNullOrWhiteSpace())
+            {
+                throw new FormatException("IoT hub service hostname to connect to must be specified.");
+            }
+
+            // Gateway host name
+            if (GatewayHostName.IsNullOrWhiteSpace())
```

Comment:
Why is this now mandatory? Users don't need a gateway to connect if they want to connect to IoT hub directly

Reply:
```csharp
        /// <summary>
        /// The host service that this client connects to.
        /// This can either be the IoT hub name or a gateway service name.
        /// </summary>
        string GatewayHostName { get; }
`

Reply:
```csharp
            IsUsingGateway = !iotHubConnectionString.GatewayHostName.IsNullOrWhiteSpace();
            GatewayHostName = IsUsingGateway
                ? iotHubConnectionString.GatewayHostName
                : IotHubHostName;
```

Reply:
This seems a bit confusing to me. If the user didn't pass in a gatewayhostname in their connection string, why would the gatewayhostname be anything but null in our SDK?

Reply:
That's fair. I think the name "gateway" is what's misleading there. We essentially need 3 fields - IotHubHostName, GatewayHostName and a "connecting-service-endpoint" that holds the value for this expression:
```c#
connecting-service-endpoint = iotHubConnectionString.GatewayHostName.IsNullOrWhiteSpace()
                ? IotHubHostName
                : iotHubConnectionString.GatewayHostName;
```

Reply:
Any suggestions for "connecting-service-endpoint"?

Reply:
"connecting-service-endpoint" as in the server that this client _directly_ connects to? "hostname" seems like a good enough name to me